### PR TITLE
[wip] Add layout for ergodox_ez with mac/german setting

### PR DESCRIPTION
**Pull request by neuhalje**
Add new erogdox_ez layout for German and MacOS/Windows.

layout

legend

### CODE REVIEW

**Comment by jackhumbert** [comment]
Nice! Unfortunately we aren't able to add submodules for keymaps like this - if you'd like to contribute yours, we'd be happy to accept it as the files themselves - we'd just encourage you to make updates as PRs sporadically.

**Comment by neuhalje** [comment]
No problem! As soon as I get a few spare minutes. I'll keep this PR open until then

**Comment by neuhalje** [comment]
* move source in qmk project
* add windows support

I "force added" the layout PNGs because without them there would be no layout documentation.

**Comment by jackhumbert** [comment]
Cool! We're trying to keep the repo size down though - if you could host the images externally, that'd be awesome!

**Comment by drashna** [comment]
We'd still like to merge this, but we do need you to remove the images.  

You can always link them in your readme.md. Hosting them on Imgur or the like would be fantastic.

**Comment by neuhalje** [comment]
I'll rebase and host on imgur (but need some time to do so, please leave the PR open)

**Comment by drashna** [comment]
Just wanted to touch bases with you, and see if you'd been able to take care of this?

**Comment by drashna** [comment]
Any updates on this?

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef CONFIG_USER_H
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+#define CONFIG_USER_H
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h]:
```suggestion
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+#endif
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h]:
```suggestion
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+#define NORMAL_MODE 0
+#define INSERT_MODE 1
+#define SYMB 2
+#define MOUSE 3
+
+#include "../../config.h"
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/config.h]: This line is not actually needed, and should be removed. 

```suggestion
```

Specifically, the build script automatically includes all of the config.h files, so this is not necessary.

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 5]
```diff
@@ -0,0 +1,357 @@
+#include "ergodox_ez.h"
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: Could you replace this:
```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+#include "debug.h"
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: Could you remove all of these additional includes?  

```suggestion
```

They get included automatically, at various points in the built process, so they are redundant and not necessary here.

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+#include "action_layer.h"
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: Could you remove all of these additional includes?  

```suggestion
```

They get included automatically, at various points in the built process, so they are redundant and not necessary here.

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+
+
+#include "keymap_german.h"
+
+#include "keymap_nordic.h"
+
+
+
+enum custom_keycodes {
+  PLACEHOLDER = SAFE_RANGE, // can always be here
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: You don't actually need `PLACEHOLDER`, and could use: 

```suggestion
  VRSN = SAFE_RANGE, // can always be here
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+  EPRM,
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: This isn't actually needed anymore. `EEP_RST` does this, and is a core keycode now.

```suggestion
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+#define WIN_EMPTY1  8
+#define WIN_EMPTY2  9
+
+#define ALL_MOUSE  10
+
+#define RGBLIGHT_COLOR_LAYER_WIN 0xff,0x00,0x00
+#define RGBLIGHT_COLOR_LAYER_MAC 0x00,0xff,0x00
+
+
+#define EMPTY_LAYER KEYMAP(KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO)
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
#define EMPTY_LAYER LAYOUT_ergodox(KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO)
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+  return (layer >= MACOS_ALPHA) && (layer <= MACOS_EMPTY2);
+}
+
+bool is_win_layer(uint8_t layer){
+  return (layer >= WIN_ALPHA) && (layer <= WIN_EMPTY2);
+}
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+  [MACOS_ALPHA] = KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
  [MACOS_ALPHA] = LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+    MAC_DE_RIGHT_SQUARE_BRACE,KC_6,KC_7,KC_8,KC_9,KC_0,DE_LESS,
+    MAC_DE_RIGHT_CURLY_BRACE,DE_Z,KC_U,KC_I,KC_O,KC_P,MAC_DE_SLASH,
+    KC_H,KC_J,KC_K,KC_L,LT(ALL_MOUSE,MAC_DE_QUESTION),OSL(MACOS_UMLAUT),NO_RPRN,
+    KC_N,KC_M,KC_COMMA,KC_DOT,NO_MINS,KC_RSHIFT,
+    KC_UP,KC_DOWN,KC_NO,KC_NONUS_HASH,KC_RALT,
+    // BLOCK
+    KC_NO,KC_NO,KC_PGUP,KC_PGDOWN,KC_ENTER,KC_SPACE
+  ),
+
+  [MACOS_SPECIAL] =  KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
  [MACOS_SPECIAL] =  LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+    KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NONUS_HASH,MAC_DE_SLASH,
+    KC_LEFT,KC_DOWN,KC_UP,KC_RIGHT,MAC_DE_PIPE,KC_KP_PLUS,
+    KC_NO,TO(MACOS_ALPHA),KC_NO,KC_NO,KC_NO,MAC_DE_BACKSLASH,KC_KP_MINUS,
+    KC_NO,KC_NO,KC_NO,MAC_DE_TILDE,MAC_DE_GRAVE,
+    // BLOCK
+    RGB_TOG,RGB_SLD,KC_NO,KC_NO,RGB_HUD,RGB_HUI
+  ),
+
+
+  [MACOS_UMLAUT] = KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
  [MACOS_UMLAUT] = LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+  [MACOS_EMPTY2] = EMPTY_LAYER,
+
+
+//
+//
+// Windows
+//
+//
+
+    [WIN_ALPHA] = KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
    [WIN_ALPHA] = LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+      DE_RBRC,KC_6,KC_7,KC_8,KC_9,KC_0,DE_LESS,
+      DE_RCBR,DE_Z,KC_U,KC_I,KC_O,KC_P,NO_SLSH,
+      KC_H,KC_J,KC_K,KC_L,LT(ALL_MOUSE,WIN_DE_QUESTION),OSL(WIN_UMLAUT),NO_RPRN,
+      KC_N,KC_M,KC_COMMA,KC_DOT,NO_MINS,KC_RSHIFT,
+      KC_UP,KC_DOWN,KC_NO,KC_NONUS_HASH,KC_RALT,
+      // BLOCK
+      KC_LALT,CTL_T(KC_ESCAPE),KC_PGUP,KC_PGDOWN,KC_ENTER,KC_SPACE
+    ),
+
+    [WIN_SPECIAL] =  KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
    [WIN_SPECIAL] =  LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+      KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NONUS_HASH,NO_SLSH,
+      KC_LEFT,KC_DOWN,KC_UP,KC_RIGHT,MAC_DE_PIPE,KC_KP_PLUS,
+      KC_NO,KC_NO,TO(MACOS_ALPHA),KC_NO,KC_NO,KC_NONUS_BSLASH,KC_KP_MINUS,
+      KC_NO,KC_NO,KC_NO,DE_TILD,WIN_DE_GRAVE,
+      // BLOCK
+      RGB_TOG,RGB_SLD,KC_NO,KC_NO,RGB_HUD,RGB_HUI
+    ),
+
+
+    [WIN_UMLAUT] = KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
    [WIN_UMLAUT] = LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+      KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_RSHIFT, // 7
+      KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,  // 5
+      // BLOCK
+      KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO // 6
+    ),
+
+    [WIN_EMPTY1] = EMPTY_LAYER,
+    [WIN_EMPTY2] = EMPTY_LAYER,
+
+  [ALL_MOUSE] = KEYMAP(
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]:
```suggestion
  [ALL_MOUSE] = LAYOUT_ergodox(
```

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+    KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,
+    KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,KC_NO,
+    KC_NO,KC_NO,KC_NO,KC_MEDIA_PLAY_PAUSE,KC_NO,KC_NO,
+    KC_NO,KC_MEDIA_PREV_TRACK,KC_MEDIA_NEXT_TRACK,KC_NO,KC_NO,KC_AUDIO_VOL_UP,KC_AUDIO_VOL_DOWN,
+    KC_AUDIO_MUTE,KC_NO,KC_NO,KC_NO,KC_NO,
+    // BLOCK
+    KC_NO,KC_NO,KC_NO,KC_WWW_BACK),
+};
+
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: Could you remove the `fn_actions` block?  All of it's features have been added as keycodes, and been deprecated in favor of `process_record_user`.

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+  [1] = ACTION_LAYER_TAP_TOGGLE(1)
+};
+
+// leaving this in place for compatibilty with old keymaps cloned and re-compiled.
+const macro_t *action_get_macro(keyrecord_t *record, uint8_t id, uint8_t opt)
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: Could you remove the `action_get_macro` function?  It's been deprecated in favor of `process_record_user`.

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+};
+
+void matrix_init_user(void) {
+  setup_backlight(MACOS_ALPHA);
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  switch (keycode) {
+    // dynamically generate these.
+    case EPRM:
```
Comment (drashna) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/keymap.c]: This case is no longer needed.

**Comment by drashna** [comment]
Any update on this?

**Comment by drashna** [comment]
Is there any updates on this? 

Mostly, the images need to be removed.  The rest are more cosmentic than anything and aren't strictly necessary. 

If you could, please remove the images. 

Otherwise, since it has been a while since there have been any updates here.

**Comment by neuhalje** [comment]
Implemented the CRs -- still need to test, so hold your horses :-)

**Comment by drashna** [comment]
Awesome, thank you for updating the PR.

There are a few more review commends that I would like see addressed (like the initial include), to keep things up to date with our coding conventions.   But these are more minor things. 

Otherwise, this is looking good!

File: keyboards/ergodox_ez/keymaps/mac_win_german_programmer/.gitignore [hunk 21]
```diff
@@ -0,0 +1 @@
+.DS_Store
```
Comment (fauxpark) [inline keyboards/ergodox_ez/keymaps/mac_win_german_programmer/.gitignore]: This file can be removed, the top-level gitignore now has the correct case to ignore these :)

**Comment by drashna** [comment]
Any update on this?

**Comment by neuhalje** [comment]
It has been largely reworked ... right now stuck in my $DAYJOB ...  will continue on that :-)

**Comment by drashna** [comment]
lol, okay. No worries.

**Comment by tzarc** [comment]
This PR hasn't seen progress in quite some time. As such, I'm removing the labels that prevent stalebot from auto-closing this PR. You're free to add commits or comments to keep it alive, but for now there's no reason to keep it open if it's going to sit here inactive.

The requirements for PRs have changed since it was raised, so if you'd like to continue the process of getting this into QMK, please refer to the [PR Checklist](https://docs.qmk.fm/#/pr_checklist) and make the required changes in your next batch of commits. Feel free to reach out to any of the collaborators if you're unsure of things, or even better, ping one of the collaborators on [Discord](https://discord.gg/Uq7gcHh) and we can take it from there. Thanks!

**Pull request closed without merging**